File: java-devildyw/src/main/java/com/dyw/leetcode/algorithm/dp/L53最大子数组和.java
package com.dyw.leetcode.algorithm.dp;

/**
 * @author Devil
 * @since 2022-11-13-16:26
 */
@SuppressWarnings("all")
public class L53最大子数组和 {

    public int maxSubArray(int[] nums) {
         int ans = Integer.MIN_VALUE;
         int sum = 0;

        for (int i = 0; i < nums.length; i++) {
            if (sum<=0){
                sum = nums[i];
            }else{
                sum+=nums[i];
            }
            ans = Math.max(sum,ans);

        }
        return ans;
    }



}

File: golang-devildyw/middle/L53最大子数组和.go
package main

func main() {}

func maxSubArray(nums []int) int {
	maxSum := nums[0]
	// 前缀和
	currentSum := 0

	for i := range len(nums) {
		// 如果前缀和是负数 则视为累赘丢弃 从0开始
		if currentSum > 0 {
			currentSum += nums[i]
		} else {
			currentSum = nums[i]
		}
		maxSum = max(currentSum, maxSum)
	}
	return maxSum
}
